common/service-support-cdi/src/main/java/io/helidon/common/servicesupport/cdi/CdiExtensionBase.java

```diff
@@ -0,0 +1,484 @@
+    }
+
+    /**
+     * Records a producer method defined by the application. Ignores producers
+     * with non-default qualifiers and library producers.
+     *
+     * @param ppm Producer method.
+     */
+    protected void recordProducerMethod(ProcessProducerMethod<?, ?> ppm) {
+        recordProducerMember("recordProducerMethod", ppm.getAnnotatedProducerMethod(), ppm.getBean());
+    }
+
+    protected Map<Bean<?>, AnnotatedMember<?>> producers() {
+        return producers;
+    }
+
+    /**
+     * Registers the service-related endpoint, after security and as CDI initializes the app scope, returning the default
+     * routing for optional use by the caller.
+     *
+     * @param adv app-scoped initialization event
+     * @param bm BeanManager
+     * @return default routing
+     */
+    protected Routing.Builder registerService(
+            @Observes @Priority(LIBRARY_BEFORE + 10) @Initialized(ApplicationScoped.class) Object adv,
+            BeanManager bm) {
+        Config config = ((Config) ConfigProvider.getConfig()).get(configPrefix);
+
+        serviceSupport = serviceSupportFactory.apply(config);
+
+        ServerCdiExtension server = bm.getExtension(ServerCdiExtension.class);
+
+        ConfigValue<String> routingNameConfig = config.get("routing").asString();
+        Routing.Builder defaultRouting = server.serverRoutingBuilder();
+
+        Routing.Builder endpointRouting = defaultRouting;
+
+        if (routingNameConfig.isPresent()) {
+            String routingName = routingNameConfig.get();
+            // support for overriding this back to default routing using config
+            if (!"@default".equals(routingName)) {
+                endpointRouting = server.serverNamedRoutingBuilder(routingName);
+            }
+        }
+
+        serviceSupport.configureEndpoint(endpointRouting);
+
+        return defaultRouting;
+    }
+
+    protected T serviceSupport() {
+        return serviceSupport;
+    }
+
+    protected void recordProducerMember(String logPrefix, AnnotatedMember<?> member, Bean<?> bean) {
+        logger.log(Level.FINE, () -> logPrefix + " " + bean.getBeanClass());
+        if (!ownProducer.equals(bean.getBeanClass())) {
+            Set<Class<? extends Annotation>> siteAnnotationTypes = new HashSet<>();
+
+            for (Annotation memberAnnotation : member.getAnnotations()) {
+                Class<? extends Annotation> memberAnnotationType = memberAnnotation.annotationType();
+                if (annotations.contains(memberAnnotationType)) {
+                    siteAnnotationTypes.add(memberAnnotationType);
+                }
+            }
+            if (!siteAnnotationTypes.isEmpty()) {
+                Optional<Class<? extends Annotation>> hasQualifier
+                        = siteAnnotationTypes
+                        .stream()
+                        .filter(annotationType -> annotationType.isAnnotationPresent(Qualifier.class))
```

Comment:
You're saying that something like `bean.getQualifiers().isEmpty()` would be better (instead of processing the site `AnnotationType`s looking for any with `@Qualifier`?

Reply:
Yes, although it's worth noting as well that every bean has, implicitly or explicitly, the `@Any` qualifier and the `@Default` qualifier (unless someone has explicitly qualified it with some other qualifier).  So as a CDI person I start to get uncomfortable when you speak of some bean somewhere not having a qualifier.

Reply:
Revise and extend... Looks like this is better (because `bean.getQualifiers()` returns a collection which contains literals for the annotations, not the classes for them):
```
if (bean.getQualifiers().stream()
                    .filter(Default.class::isInstance)
                    .findFirst()
                    .isPresent()) {
                producers.put(bean, member);
            }
```

Reply:
Yes.  

In case this sort of thing becomes an issue anywhere else, `InjectionPoint`s also have qualifiers available via a `getQualifiers()` method.  In general, an injection point gets `@Default` unless otherwise specified, and a bean gets `@Default` if there is no other qualifier other than `@Named` on it, and always (unless you're really determined!) gets `@Any` as well.  Then matching a bean to an injection point happens via subsets: the injection point has to have a subset of a bean's qualifiers to match.

I mention all of this because people can declare any old annotation they want (within some reason) as a qualifier by just…telling CDI that it is one.  Then those annotations can show up in these `getQualifiers()` collections, and won't be "findable" by rooting around in an annotated element's annotations anywhere.